lib/pbio/drv/block_device/block_device_flash_stm32.c

```diff
@@ -28,34 +30,112 @@ uint32_t pbdrv_block_device_get_size(void) {
 void pbdrv_block_device_init(void) {
 }
 
+static void (*_callback)(void);
 void pbdrv_block_device_set_callback(void (*callback)(void)) {
+    _callback = callback;
 }
 
+extern uint8_t _pbdrv_block_device_storage_start[];
+
 PT_THREAD(block_device_read(struct pt *pt, uint32_t offset, uint8_t *buffer, uint32_t size, pbio_error_t *err)) {
 
     PT_BEGIN(pt);
 
-    // TODO: Read from user flash. For now this just loads
-    // the mpy file appended to the firmware to keep Pybricks Code working.
-    extern uint32_t _pb_user_mpy_size;
-    if (_pb_user_mpy_size + 8 > pbdrv_block_device_get_size()) {
-        *err = PBIO_ERROR_FAILED;
+    // Exit on invalid size.
+    if (size == 0 || offset + size > pbdrv_block_device_get_size()) {
+        *err = PBIO_ERROR_INVALID_ARG;
         PT_EXIT(pt);
     }
-    extern uint8_t _pb_user_mpy_data[];
-    pbio_set_uint32_le(buffer, _pb_user_mpy_size + 8);
-    pbio_set_uint32_le(&buffer[4], _pb_user_mpy_size);
-    memcpy(&buffer[8], _pb_user_mpy_data, _pb_user_mpy_size);
+
+    // Copy requested data to RAM.
+    memcpy(buffer, _pbdrv_block_device_storage_start + offset, size);
     *err = PBIO_SUCCESS;
 
     PT_END(pt);
 }
 
-PT_THREAD(block_device_store(struct pt *pt, uint8_t *buffer, uint32_t size, pbio_error_t *err)) {
+typedef union _double_word_t {
+    uint8_t data[8];
+    uint64_t dword;
+} double_word_t;
+
+static pbio_error_t block_device_erase_and_write(uint8_t *buffer, uint32_t size) {
+
+    static const uint32_t base_address = (uint32_t)(&_pbdrv_block_device_storage_start[0]);
+
+    // Exit on invalid size.
+    if (size == 0 || size > pbdrv_block_device_get_size()) {
+        return PBIO_ERROR_INVALID_ARG;
+    }
+
+    // Unlock flash for writing.
+    HAL_StatusTypeDef hal_err = HAL_FLASH_Unlock();
+    if (hal_err != HAL_OK) {
+        return PBIO_ERROR_IO;
+    }
+
+    // Erase only as much as we need.
+    FLASH_EraseInitTypeDef erase_init = {
+        #if defined(STM32F0)
+        .PageAddress = base_address,
+        #elif defined(STM32L4)
+        .Banks = FLASH_BANK_1,
+        .Page = (FLASH_SIZE - (pbdrv_block_device_get_size())) / FLASH_PAGE_SIZE,
+        #else
+        #warning "Unsupported target."
+        #endif
+        .NbPages = size / FLASH_PAGE_SIZE + (size % FLASH_PAGE_SIZE != 0),
```

Comment:
```suggestion
        .NbPages = (size + (FLASH_PAGE_SIZE - 1) / FLASH_PAGE_SIZE,
```

This is a more idiomatic way to round up while dividing in C.